### PR TITLE
Mad Tenscripting

### PR DESCRIPTION
**Pull request by geralddejong**
Some big changes in how tenscript is structured

### CODE REVIEW

File: src/build/growth.rs [hunk 1]
```diff
@@ -102,108 +110,134 @@ impl Growth {
...
-    fn complete_shaper(&self, fabric: &mut Fabric, Shaper { interval, alpha_face, omega_face }: &Shaper) {
-        let (alpha, omega) = (fabric.face(*alpha_face), fabric.face(*omega_face));
+    fn attach_shapers_for(&self, fabric: &mut Fabric, shaper_spec: &ShaperSpec) -> Vec<Shaper> {
+        let mut shapers: Vec<Shaper> = vec![];
+        match shaper_spec {
+            ShaperSpec::Join { mark_name } => {
+                let faces = self.marked_faces(mark_name);
+                let joints = self.marked_middle_joints(fabric, &faces);
+                match joints.len() {
+                    2 => {
```
Comment (mitchelljustin) [inline src/build/growth.rs:194]: Are you going to add more match arms in the future?

File: src/build/growth.rs [hunk 2]
```diff
@@ -102,108 +110,134 @@ impl Growth {
...
+                        let interval = fabric.create_interval(joints[0], joints[1], Link::Pull { ideal: 0.3 });
+                        shapers.push(Shaper { interval, alpha_face: faces[0], omega_face: faces[1], join: true })
+                    }
+                    _ => unimplemented!()
+                }
```
Comment (mitchelljustin) [inline src/build/growth.rs:199]:
```suggestion
                match joints.as_slice() {
                    &[alpha_index, omega_index] => {
                        let interval = fabric.create_interval(alpha_index, omega_index, Link::Pull { ideal: 0.3 });
                        shapers.push(Shaper { interval, alpha_face: faces[0], omega_face: faces[1], join: true })
                    }
                    _ => unimplemented!()
                }
```

File: src/build/growth.rs [hunk 3]
```diff
@@ -228,8 +262,27 @@ impl Growth {
...
+        fabric.remove_face(alpha_id);
+        fabric.remove_face(omega_id);
+    }
+
+    fn branch_pairs(nodes: &[TenscriptNode]) -> Vec<(FaceName, &TenscriptNode)> {
+        nodes
+            .iter()
+            .map(|face_node| {
+                let Face { face_name, node } = face_node else {
+                    panic!("Branch may only contain Face nodes");
```
Comment (mitchelljustin) [inline src/build/growth.rs:274]: U sure you don't wanna return a Result instead? Seems hacky/lazy to panic

File: src/build/plan_runner.rs [hunk 4]
```diff
@@ -49,7 +27,7 @@ pub struct PlanRunner {
 impl Default for PlanRunner {
     fn default() -> Self {
         Self {
-            growth: Growth::new(parse(CODE).unwrap()),
+            growth: Growth::new(fabric_plan("Halo by Crane")),
```
Comment (mitchelljustin) [inline src/build/plan_runner.rs:30]: Very nice

File: src/build/tenscript/bootstrap.scm [hunk 5]
```diff
@@ -0,0 +1,73 @@
+;;;Seed
```
Comment (mitchelljustin) [inline src/build/tenscript/bootstrap.scm:1]: A bit weird to put the names in comments but you know this. We'll get to including the in the (fabric ..) later

File: src/build/tenscript/expression.rs [hunk 6]
```diff
@@ -101,25 +103,24 @@ impl Parser {
...
-            Paren('(') => self.list(),
-            Ident(name) => Ok(Expression::Ident(name)),
-            Float(value) => Ok(Expression::Float(value)),
-            Integer(value) => Ok(Expression::Integer(value)),
-            Percent(value) => Ok(Expression::Percent(value)),
             Atom(value) => Ok(Expression::Atom(value)),
-            Token::String(value) => Ok(Expression::String(value)),
+            FloatingPoint(value) => Ok(Expression::FloatingPoint(value)),
+            Identifier(name) => Ok(Expression::Identifier(name)),
+            Parenthesis('(') => self.list(),
```
Comment (mitchelljustin) [inline src/build/tenscript/expression.rs:109]: He don't like short names for things huh. No parens or idents

File: src/build/tenscript/mod.rs [hunk 7]
```diff
@@ -51,54 +52,105 @@ impl Spin {
...
-        branch: Option<Box<TenscriptNode>>,
+        post_growth_node: Option<Box<TenscriptNode>>,
     },
     Mark {
-        face_name: FaceName,
         mark_name: String,
     },
     Branch {
-        subtrees: Vec<TenscriptNode>,
+        face_nodes: Vec<TenscriptNode>,
```
Comment (mitchelljustin) [inline src/build/tenscript/mod.rs:68]: Good name change

File: src/build/growth.rs (hunk 2 above)
Comment (geralddejong) [reply src/build/growth.rs:199]: Nice!

File: src/build/tenscript/bootstrap.scm (hunk 5 above)
Comment (geralddejong) [reply src/build/tenscript/bootstrap.scm:1]: Yes when you redo the parser.

File: src/build/tenscript/expression.rs (hunk 6 above)
Comment (geralddejong) [reply src/build/tenscript/expression.rs:109]: Too nerdy

File: src/build/tenscript/mod.rs [hunk 8]
```diff
@@ -51,54 +52,105 @@ impl Spin {
...
 pub struct FabricPlan {
-    pub name: Option<String>,
     pub surface: Option<SurfaceCharacterSpec>,
-    pub features: Features,
     pub build_phase: BuildPhase,
     pub shape_phase: ShapePhase,
 }
+
+fn bootstrap() -> HashMap<&'static str, &'static str> {
+    include_str!("bootstrap.scm")
```
Comment (mitchelljustin) [inline src/build/tenscript/mod.rs:122]: Pretty cool that Rust can do this, &'static strs from a text file

File: src/build/tenscript/parser.rs [hunk 9]
```diff
@@ -20,11 +19,10 @@ impl Display for ParseError {
...
-    Mismatch { rule: &'static str, expression: Expression, expected: &'static str },
-    BadCall { context: &'static str, expected: &'static str, expression: Expression },
+    Mismatch { expected: &'static str, expression: Expression },
+    BadCall { expected: &'static str, expression: Expression },
     TypeError { expected: &'static str, expression: Expression },
     AlreadyDefined { property: &'static str, expression: Expression },
-    IllegalRepetition { kind: &'static str, value: String },
     MultipleBranches,
     IllegalCall { context: &'static str, expression: Expression },
     Unknown,
```
Comment (mitchelljustin) [inline src/build/tenscript/parser.rs:28]: This should probably be cleaned up even more (by me) in the future

File: src/build/tenscript/parser.rs [hunk 10]
```diff
@@ -42,76 +40,62 @@ pub fn parse(source: &str) -> Result<FabricPlan, Error> {
...
+//                 };
+//                 match name.as_str() {
+//                     $(
+//                         $name => $enum_val,
+//                     )+
+//                     _ => return Err($crate::build::tenscript::parser::ErrorKind::TypeError { expected, expression: $value.clone() })
+//                 }
+//             }
+//         }
+//     }
```
Comment (mitchelljustin) [inline src/build/tenscript/parser.rs:58]: Nuke it!

File: src/build/tenscript/mod.rs (hunk 8 above)
Comment (geralddejong) [reply src/build/tenscript/mod.rs:122]: Awe's what I've got.  So many good decisions in this language.

File: src/build/tenscript/parser.rs [hunk 11]
```diff
@@ -42,76 +40,62 @@ pub fn parse(source: &str) -> Result<FabricPlan, Error> {
...
 
 struct Call<'a> {
     head: &'a str,
     tail: &'a [Expression],
 }
 
-fn expect_call<'a>(rule: &'static str, expression: &'a Expression) -> Result<Call<'a>, ErrorKind> {
+fn expect_call(expression: &Expression) -> Result<Call, ErrorKind> {
```
Comment (mitchelljustin) [inline src/build/tenscript/parser.rs:65]: You don't want to pass in the rule anymore? I've always found that quite useful in debugging parser errors

File: src/build/tenscript/parser.rs (hunk 10 above)
Comment (geralddejong) [reply src/build/tenscript/parser.rs:58]: Yeh, forgot.

File: src/build/tenscript/parser.rs (hunk 11 above)
Comment (geralddejong) [reply src/build/tenscript/parser.rs:65]: Something about the static lifetime kept causing a mild-disapproval.

File: src/build/tenscript/parser.rs [hunk 12]
```diff
@@ -127,20 +111,21 @@ fn fabric_plan(expression: &Expression) -> Result<FabricPlan, ErrorKind> {
...
+                    "right-left" => Seed::RightLeft,
+                    _ => {
+                        return Err(BadCall { expected: "(seed <left | right | left-right | right-left>)", expression: expression.clone() });
+                    }
                 };
-                let seed_type = expect_enum!(value, {
-                        "left" => Spin::Left,
-                        "right" => Spin::Right,
-                    });
-                build_phase.seed = Some(seed_type);
```
Comment (mitchelljustin) [inline src/build/tenscript/parser.rs:143]: Honestly clearer than what I did

File: src/build/tenscript/bootstrap.scm (hunk 5 above)
Comment (mitchelljustin) [reply src/build/tenscript/bootstrap.scm:1]: ;)

File: src/fabric/twist.rs [hunk 13]
```diff
@@ -50,10 +48,10 @@ impl Fabric {
             self.create_interval(alpha, omega, Link::Pull { ideal: ROOT3 * scale });
         }
         if let Some(id) = face_id { self.faces_to_loop(id, a_minus_face) }
-        [(Aneg, a_minus_face), (Apos, a_plus_face)]
+        vec![(Aneg, a_minus_face), (Apos, a_plus_face)]
```
Comment (mitchelljustin) [inline src/fabric/twist.rs:51]: How come?

**Review by mitchelljustin — APPROVED** [review]

File: src/build/tenscript/parser.rs (hunk 11 above)
Comment (mitchelljustin) [reply src/build/tenscript/parser.rs:65]: Weird haha

File: src/build/tenscript/mod.rs [hunk 14]
```diff
@@ -51,54 +52,105 @@ impl Spin {
...
+mod tests {
+    use crate::build::tenscript::{bootstrap, parser};
+
+    #[test]
+    fn parse() {
+        let map = bootstrap();
+        for (name, code) in map.into_iter() {
+            match parser::parse(code) {
+                Ok(_) => println!("[{name}] Good plan!"),
+                Err(error) => println!("[{name}] Error: {error:?}"),
```
Comment (mitchelljustin) [inline src/build/tenscript/mod.rs:152]: Should panic here or something..

**Pull request merged**